### PR TITLE
feat: Scanner rule for unfulfilled promises

### LINKED ISSUE #1477: Unfulfilled promises are scanned for and reported as flaws

### PR DESCRIPTION
**Pull request by dividedmind**
Fixes #1477

### CODE REVIEW

File: packages/scanner/doc/rules/unfulfilled-promise.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+AppMap with the resolved value and total elapsed run time.
+
+However, if a promise remains unfulfilled when the recording ends (for example when a HTTP response
+is sent or a test case finishes) this pending state is never updated.
+
+Most of the time it means some asynchronous computation has been started and its results not awaited
+for and therefore not collected; such computations should be removed or the results awaited.
+
+In some cases this is an intended behaviour; for example, a background book-keeping or cleanup
+process might be start that doesn't directly impact the results of the operation that triggered it.
```
Comment (brikelly) [inline packages/scanner/doc/rules/unfulfilled-promise.md:29]:
```suggestion
process might be started but doesn't directly impact the results of the operation that triggered it.
```

File: packages/scanner/doc/rules/unfulfilled-promise.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+However, testing assumes clean and known environment at the beginning of each case so in testing
```
Comment (brikelly) [inline packages/scanner/doc/rules/unfulfilled-promise.md:30]: This line reads weird, not sure how to fix it. Can you re-phrase it please?

**Review by brikelly — CHANGES_REQUESTED** [review]

File: packages/scanner/doc/rules/unfulfilled-promise.md (hunk 1 above)
Comment (dividedmind) [reply packages/scanner/doc/rules/unfulfilled-promise.md:29]: Should be "started, that" actually, thanks.

File: packages/scanner/doc/rules/unfulfilled-promise.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+Most of the time it means some asynchronous computation has been started and its results not awaited
+for and therefore not collected; such computations should be removed or the results awaited.
+
+In some cases this is an intended behaviour; for example, a background book-keeping or cleanup
+process might be started which doesn't directly impact the results of the operation that triggered
+it. Note that during testing such processes should still run to completion before a test case
+finishes (or else be suppressed altogether); otherwise there might be unintended interference
+between test cases leading to brittle tests. Each test case should start from a known, clean and
+quiescent environment.
```
Comment (dividedmind) [inline packages/scanner/doc/rules/unfulfilled-promise.md:33]: @brikelly is this any better?

**Review by ahtrotta — APPROVED** [review]

**Review by brikelly — APPROVED** [review]
LGTM!

**Pull request merged**